File: doc-examples/example-java/src/main/java/example/UploadController.java
package example;

import io.micronaut.http.HttpResponse;
import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Post;
import io.micronaut.http.multipart.CompletedFileUpload;
import io.micronaut.objectstorage.aws.AwsS3Operations;
import io.micronaut.objectstorage.request.UploadRequest;
import io.micronaut.objectstorage.response.UploadResponse;
import software.amazon.awssdk.services.s3.model.ObjectCannedACL;
import software.amazon.awssdk.services.s3.model.PutObjectResponse;

@Controller
public class UploadController {

    private final AwsS3Operations objectStorage;

    public UploadController(AwsS3Operations objectStorage) {
        this.objectStorage = objectStorage;
    }

    @Post(consumes = MediaType.MULTIPART_FORM_DATA, produces = MediaType.TEXT_PLAIN)
    public HttpResponse<String> upload(CompletedFileUpload fileUpload) {
        UploadRequest objectStorageUpload = UploadRequest.fromCompletedFileUpload(fileUpload);
        UploadResponse<PutObjectResponse> response = objectStorage.upload(objectStorageUpload, builder -> {
            builder.acl(ObjectCannedACL.PUBLIC_READ);
        });

        return HttpResponse.created(response.getNativeResponse().eTag());
    }
}

File: doc-examples/example-python/src/main/python/example/UploadController.py
from micronaut.http import HttpResponse, MediaType
from micronaut.http.annotation import Controller, Post
from micronaut.http.multipart import CompletedFileUpload, StreamingFileUpload
from micronaut.objectstorage.aws import AwsS3Operations
from micronaut.objectstorage.request import UploadRequest
from software.amazon.awssdk.services.s3.model import ObjectCannedACL


# tag::beginclass[]
@Controller
class UploadController:

    def __init__(self, object_storage: AwsS3Operations):
        self.object_storage = object_storage
# end::beginclass[]

    @Post(consumes=MediaType.MULTIPART_FORM_DATA, produces=MediaType.TEXT_PLAIN)
    def upload(self, file_upload: CompletedFileUpload) -> HttpResponse[str]:
        object_storage_upload = UploadRequest.fromCompletedFileUpload(file_upload)
# tag::consumer[]
        response = self.object_storage.upload(object_storage_upload, lambda builder: builder.acl(ObjectCannedACL.PUBLIC_READ))
# end::consumer[]

        return HttpResponse \
            .created(response.getKey()) \
            .header("ETag", response.getNativeResponse().eTag())

# tag::streaming[]
    @Post(uri="/stream", consumes=MediaType.MULTIPART_FORM_DATA, produces=MediaType.TEXT_PLAIN)
    def streaming_upload(self, file_upload: StreamingFileUpload) -> HttpResponse[str]:
        object_storage_upload = UploadRequest.fromStreamingFileUpload(file_upload, "uploads/" + file_upload.getFilename())
        response = self.object_storage.upload(object_storage_upload)
        return HttpResponse \
            .created(response.getKey()) \
            .header("ETag", response.getNativeResponse().eTag())
# end::streaming[]

# tag::endclass[]
# end::endclass[]
